### PR TITLE
feat: Custom line items

### PR DESCRIPTION
**Pull request by olivermrbl**
Adds support for custom line items and line items with a custom price. We had all the pieces to enable this feature, they just needed some adaptation to work as expected for all scenarios. 

The changes are primarily around ensuring the correct shape of line items depending on the input in a range of workflows including:
- Create carts
- Create orders
- Add to cart
- Add to order
- Refresh cart items
- Update cart
- Update cart line item
- Complete cart

The approach taken in the PR is documented in [this RFC](https://linear.app/medusajs/document/rfc-c45ff06c71b5) and all the scenarios can be found [here](https://linear.app/medusajs/document/scenarios-4ba8fa11536f).

### CODE REVIEW

File: packages/core/core-flows/src/cart/utils/prepare-line-item-data.ts [hunk 1]
```diff
@@ -1,71 +1,111 @@
...
     throw new Error("Variant does not have a product")
   }
 
-  let compareAtUnitPrice = data.compareAtUnitPrice
+  let compareAtUnitPrice = item?.compare_at_unit_price
+
+  const isSalePrice = !!(
+    variant?.calculated_price?.calculated_price?.price_list_type ===
+    PriceListType.SALE
+  )
```
Comment (carlos-r-l-rodrigues) [inline packages/core/core-flows/src/cart/utils/prepare-line-item-data.ts]:
```suggestion
  const isSalePrice = variant?.calculated_price?.calculated_price?.price_list_type ===
    PriceListType.SALE
```

**Review by carlos-r-l-rodrigues — APPROVED** [review]

**Comment by olivermrbl** [comment]
@riqwan, would appreciate if you could also lay your eyes on this PR 🙏

File: packages/core/core-flows/src/cart/utils/prepare-line-item-data.ts [hunk 2]
```diff
@@ -1,71 +1,115 @@
...
-    unitPrice,
-    isTaxInclusive,
-    quantity,
-    metadata,
     cartId,
     taxLines,
     adjustments,
+    isCustomPrice,
+    unitPrice,
+    isTaxInclusive,
```
Comment (riqwan) [inline packages/core/core-flows/src/cart/utils/prepare-line-item-data.ts:85]: q: These are all present in the item object right? They don't seem to be changing between workflows

File: packages/modules/cart/src/migrations/Migration20241125130058.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { Migration } from '@mikro-orm/migrations';
+
+export class Migration20241125130058 extends Migration {
+
+  async up(): Promise<void> {
+    this.addSql('alter table if exists "cart_line_item" add column if not exists "is_custom_price" boolean null;');
+    this.addSql('update "cart_line_item" set "is_custom_price" = false where "is_custom_price" is null;');
+    this.addSql('alter table if exists "cart_line_item" alter column "is_custom_price" set not null;');
```
Comment (riqwan) [inline packages/modules/cart/src/migrations/Migration20241125130058.ts]: q: should we set a default here?

File: integration-tests/modules/__tests__/cart/store/cart.workflows.spec.ts [hunk 4]
```diff
@@ -780,39 +1115,490 @@ medusaIntegrationTestRunner({
             select: ["id", "region_id", "currency_code", "sales_channel_id"],
           })
 
-          await addToCartWorkflow(appContainer).run({
+          await addToCartWorkflow(appContainer).run({
```
Comment (riqwan) [inline integration-tests/modules/__tests__/cart/store/cart.workflows.spec.ts:1118]: sugg: Might be worth testing a case for adjustments and tax lines here if custom items do have them.

**Review by riqwan — APPROVED** [review]
lgtm!

File: packages/core/core-flows/src/cart/utils/prepare-line-item-data.ts (hunk 2 above)
Comment (olivermrbl) [reply packages/core/core-flows/src/cart/utils/prepare-line-item-data.ts:85]: Yeah, but they have different values depending on where this step is consumed. 

For example, in the `addToCartWorkflow`, if you pass a unit price upon adding the item, they will be:
```ts
isCustomPrice: true
unitPrice: item.unit_price
isTaxInclusive: item.is_tax_inclusive || variant.calculated_price.is_tax_inclusive
```

While as in the `refreshCartItemsWorkflow`, they will be, you can't update the item price, but the variant price might have changed, so they will be:

```ts
isCustomPrice: item.is_custom_price
unitPrice: item.is_custom_price ? item.unit_price : variant.calculated_price.is_tax_inclusive
isTaxInclusive: item.is_custom_price ? item.is_tax_inclusive: variant.calculated_price.is_tax_inclusive
```

And so on and so forth.

File: packages/modules/cart/src/migrations/Migration20241125130058.ts (hunk 3 above)
Comment (olivermrbl) [reply packages/modules/cart/src/migrations/Migration20241125130058.ts]: Yeah, that was the intention. Nice catch!

File: integration-tests/modules/__tests__/cart/store/cart.workflows.spec.ts (hunk 4 above)
Comment (olivermrbl) [reply integration-tests/modules/__tests__/cart/store/cart.workflows.spec.ts:1118]: The first iteration of custom line items doesn't support tax lines and adjustments 👍

**Pull request merged**